onnxruntime/core/graph/graph_utils.cc

```diff
@@ -752,40 +752,53 @@ bool FindPath(Graph& graph, const Node& node, bool is_input_edge, const std::vec
   return true;
 }
 
-bool RemoveNodesWithOneOutputBottomUp(Graph& graph, const Node& start_node) {
+bool RemoveNodesWithOneOutputBottomUp(Graph& graph, const Node& start_node, bool force) {
   std::queue<const Node*> q;
-  std::vector<NodeIndex> nodes_to_remove;
+  std::unordered_set<NodeIndex> removed_nodes;
   q.push(&start_node);
   // From the current node, remove nodes bottom-up util it reaches a node with multiple outputs/graph output.
   while (q.size() != 0) {
     const Node& cur_node = *(q.front());
     q.pop();
+
+    if (removed_nodes.find(cur_node.Index()) != removed_nodes.end()) {
+      continue;
+    }
     // Each eligible node in the subgraph must have less than one output edge and no output should be
     // the graph output
     if (cur_node.GetOutputEdgesCount() > 1 || !graph.GetNodeOutputsInGraphOutputs(cur_node).empty()) {
       continue;
     }
-    nodes_to_remove.push_back(cur_node.Index());
+
     // push the parents of current node to the queue.
     for (unsigned int i = 0; i < cur_node.InputDefs().size(); ++i) {
       const std::string& input_name = GetNodeInputName(cur_node, i);
       if (IsInitializer(graph, input_name, true) || IsGraphInput(graph, cur_node.InputDefs()[i])) {
         // skip initializers and graph inputs
         continue;
       }
-      q.push(GetInputNode(cur_node, i));
+      const Node* child_node = GetInputNode(cur_node, i);
+      std::unordered_set<NodeIndex>::const_iterator it = removed_nodes.find(child_node->Index());
+      if (it == removed_nodes.end()) {
```

Comment:
nit: might not need check here since it will be checked after poping from queue.